OptimizelySDK/Odp/OdpEventManager.cs

```diff
@@ -0,0 +1,448 @@
+            {
+                return;
+            }
+
+            _logger.Log(LogLevel.DEBUG, "Stop requested.");
+
+            _flushIntervalCancellation.Cancel();
+            try
+            {
+                _flushQueueRegularly.ConfigureAwait(false).GetAwaiter().GetResult();
+            }
+            catch (OperationCanceledException)
+            {
+                // exception raised by design from .Cancel()
+                _logger.Log(LogLevel.DEBUG, "Cancel requested successfully.");
+            }
+            catch (Exception ex)
+            {
+                // other exceptions should be ignored
+                _logger.Log(LogLevel.ERROR, ex.Message);
+            }
+            finally
+            {
+                _flushIntervalCancellation.Dispose();
+            }
+
+            // one final time
+            FlushQueue();
+
+            lock (lockObject)
+            {
+                CurrentState = ExecutionState.Stopped;
+            }
+
+            _logger.Log(LogLevel.DEBUG, $"Stopped. Queue Count: {_queue.Count}.");
+        }
+
+        /// <summary>
+        /// Associate a full-stack userid with an established VUID
+        /// </summary>
+        /// <param name="userId">Full-stack User ID</param>
+        public void IdentifyUser(string userId)
+        {
+            if (!IsOdpConfigurationReady())
+            {
+                return;
+            }
+
+            var identifiers = new Dictionary<string, string>
+            {
+                {
+                    OdpUserKeyType.FS_USER_ID.ToString(), userId
+                },
+            };
+
+            var odpEvent = new OdpEvent(TYPE, "identified", identifiers);
+            SendEvent(odpEvent);
+        }
+
+        /// <summary>
+        /// Send an event to ODP
+        /// </summary>
+        /// <param name="odpEvent">ODP Event to forward</param>
+        public void SendEvent(OdpEvent odpEvent)
+        {
+            if (!IsOdpConfigurationReady())
+            {
+                return;
+            }
+
+            if (CurrentState == ExecutionState.Stopped)
+            {
+                _logger.Log(LogLevel.WARN, "ODP is not enabled.");
+                return;
+            }
+
+            if (InvalidDataFound(odpEvent.Data))
+            {
+                _logger.Log(LogLevel.ERROR, "ODP data is not valid.");
+                return;
+            }
+
+            odpEvent.Data = AugmentCommonData(odpEvent.Data);
+            Enqueue(odpEvent);
+        }
+
+        /// <summary>
+        /// Add a new ODP event to the queue
+        /// </summary>
+        /// <param name="odpEvent"></param>
+        private void Enqueue(OdpEvent odpEvent)
+        {
+            if (_queue.Count >= _queueSize)
+            {
+                _logger.Log(LogLevel.WARN, $"ODP event send failed (queueSize = {_queue.Count}).");
+                return;
+            }
+
+            _queue.Enqueue(odpEvent);
+
+            ProcessQueue();
```

Comment:
Do we want a blocking call for ProcessQueue for all event send? If batch is being processed, then this will block on lock until flush completed.